README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# weightedGCM

<!-- badges: start -->
<!-- badges: end -->

The package weightedGCM contains two functions implementing the two
versions of the Weighted Generalised Covariance Measure (WGCM)
conditional independence test described in Scheidegger, Hoerrmann and
Buehlmann (2021) “The Weighted Generalised Covariance Measure”
\<arXiv:2111.04361>. It is a generalisation of the Generalised
Covariance Measure (GCM) implemented in the package
‘GeneralisedCovarianceMeasure’ by Jonas Peters and Rajen D. Shah based
on Shah and Peters (2020) “The Hardness of Conditional Independence
Testing and the Generalised Covariance Measure” \<arXiv:1804.07203>.

-   *wgcm.fix* calculates a p-value for the null hypothesis of
    conditional independence based on the WGCM using several fixed
    weight functions.

-   *wgcm.est* calculates a p-value for the null hypothesis of
    conditional indepencence based on the WGCM using a single estimated
    weight function.

## Installation

You can install the released version of weightedGCM from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("weightedGCM")
```

## Example

We generate some data to use the two conditional independence test.

``` r
library(weightedGCM)
## Generate data
set.seed(1)
n <- 200
Z <- rnorm(n)
X <- Z + 0.3*rnorm(n)

## Y1 _||_ X | Z
Y1 <- Z + 0.3*rnorm(n)
## Y2 not _||_ X | Z
Y2 <- Z + 0.3*rnorm(n) + 0.3*X
## Y3 not _||_ X | Z
Y3 <- Z + 0.3*rnorm(n) + 0.15*X^2

## Test for conditional independence using wgcm.fix()
wgcm.fix(X, Y1, Z, regr.meth = "gam", weight.num = 7, weight.meth = "sign")
#> [1] 0.868
wgcm.fix(X, Y2, Z, regr.meth = "gam", weight.num = 7, weight.meth = "sign")
#> [1] 0.004
wgcm.fix(X, Y3, Z, regr.meth = "gam", weight.num = 7, weight.meth = "sign")
#> [1] 0.004

## Test for conditional independence using wgcm.est()
wgcm.est(X, Y1, Z, beta = 0.3, regr.meth = "gam")
#> [1] 0.2389116
wgcm.est(X, Y2, Z, beta = 0.3, regr.meth = "gam")
#> [1] 9.974068e-05
wgcm.est(X, Y3, Z, beta = 0.3, regr.meth = "gam")
#> [1] 0.004474768
```


NEWS.md
# weightedGCM 0.1.0
* First version of the package.


tests/testthat.R
library(testthat)
library(weightedGCM)

test_check("weightedGCM")


tests/testthat/test_wgcmest.R
test_that("Z = NULL yields a warning", {
  #univariate
  set.seed(1)
  n <- 200
  X <- rnorm(n)
  Y <- X + 0.3*rnorm(n)
  expect_warning(wgcm.est(X, Y, Z = NULL, beta = 0.3, regr.meth = "gam", nsim = 499),
                 "No Z specified. No weight function can be estimated. Function simply tests for vanishing correlation between X and Y.")
  #multivariate
  set.seed(1)
  X <- cbind(rnorm(n), rnorm(n))
  Y <- cbind(X[,1] + 0.3*rnorm(n), rnorm(n))
  expect_warning(wgcm.est(X, Y, Z = NULL, beta = 0.3, regr.meth = "gam", nsim = 499),
                 "No Z specified. No weight functions can be estimated. Function simply tests for vanishing correlation between components of X and Y.")

})

test_that("wgcm.est works", {
  skip_on_cran()
  ##univariate X,Y
  set.seed(1)
  n <- 200
  Z <- rnorm(n)
  Zm <- cbind(Z, Z+rnorm(n), rnorm(n), rnorm(n), rnorm(n))
  X <- Z + 0.3*rnorm(n)
  Y <- Z + 0.3*rnorm(n)
  #univariate Z
  expect_silent(wgcm.est(X, Y, Z, beta = 0.3, regr.meth = "gam", nsim = 499))
  expect_silent(wgcm.est(X, Y, Z, beta = 0.3, regr.meth = "xgboost", nsim = 499))
  #multivariate Z
  expect_silent(wgcm.est(X, Y, Zm, beta = 0.3, regr.meth = "gam", nsim = 499))
  expect_silent(wgcm.est(X, Y, Zm, beta = 0.3, regr.meth = "xgboost", nsim = 499))

  ##multivariate X, Y
  X <- cbind(X, X+rnorm(n), rnorm(n))
  Y <- cbind(Y, rnorm(n))
  #univariate Z
  expect_silent(wgcm.est(X, Y, Z, beta = 0.3, regr.meth = "gam", nsim = 499))
  expect_silent(wgcm.est(X, Y, Z, beta = 0.3, regr.meth = "xgboost", nsim = 499))
  #multivariate Z
  expect_silent(wgcm.est(X, Y, Zm, beta = 0.3, regr.meth = "gam", nsim = 499))
  expect_silent(wgcm.est(X, Y, Zm, beta = 0.3, regr.meth = "xgboost", nsim = 499))
})


tests/testthat/test_wgcmfix.R
test_that("weight_matrix() gives the right result", {
  Z <- seq(0, 1, length.out = 24)
  W1 <- cbind(rep(1,24), c(rep(-1,12), rep(1,12)))
  W2 <- cbind(rep(1,24), c(rep(-1,8), rep(1, 16)), c(rep(-1,16), rep(1,8)))
  W3 <- cbind(rep(1,24), c(rep(-1, 6), rep(1,18)),
              c(rep(-1,12), rep(1, 12)), c(rep(-1, 18), rep(1, 6)))
  W5 <- cbind(rep(1, 24), c(rep(-1, 4), rep(1, 20)), c(rep(-1, 8), rep(1, 16)),
              c(rep(-1, 12), rep(1, 12)), c(rep(-1, 16), rep(1, 8)), c(rep(-1, 20), rep(1, 4)))

  expect_equal(unname(weight_matrix(as.matrix(Z), weight.num = 1, weight.meth = "sign")), W1)
  expect_equal(unname(weight_matrix(as.matrix(Z), weight.num = 2, weight.meth = "sign")), W2)
  expect_equal(unname(weight_matrix(as.matrix(Z), weight.num = 3, weight.meth = "sign")), W3)
  expect_equal(unname(weight_matrix(as.matrix(Z), weight.num = 5, weight.meth = "sign")), W5)

  ZZ <- cbind(seq(0, 1, length.out = 6), - seq(0, 1, length.out = 6))
  W <- cbind(rep(1, 6), c(rep(-1, 2), rep(1, 4)), c(rep(-1, 4), rep(1, 2)), c(rep(1, 4), rep(-1, 2)), c(rep(1, 2), rep(-1,4)))
  expect_equal(unname(weight_matrix(as.matrix(ZZ), weight.num = 2, weight.meth= "sign")), W)
})


test_that("Z = NULL yields a warning", {
  #univariate
  set.seed(1)
  n <- 200
  X <- rnorm(n)
  Y <- X + 0.3*rnorm(n)
  expect_warning(wgcm.fix(X, Y, Z = NULL, regr.meth = "gam", weight.num = 7, weight.meth = "sign"),
                 "No Z specified. No weight functions can be calculated. Function simply tests for vanishing correlation between X and Y.")
  #multivariate
  set.seed(1)
  X <- cbind(rnorm(n), rnorm(n))
  Y <- cbind(X[,1] + 0.3*rnorm(n), rnorm(n))
  expect_warning(wgcm.fix(X, Y, Z = NULL, regr.meth = "gam", weight.num = 7, weight.meth = "sign"),
                 "No Z specified. No weight functions can be calculated. Function simply tests for vanishing correlation between components of X and Y.")

})

test_that("wgcm.fix works", {
  skip_on_cran()
  ##univariate X and Y
  set.seed(1)
  n <- 200
  Z <- rnorm(n)
  Zm <- cbind(Z, Z+rnorm(n), rnorm(n), rnorm(n), rnorm(n))
  X <- Z + 0.3*rnorm(n)
  Y <- Z + 0.3*rnorm(n)
  #weight.num = 7
  expect_silent(wgcm.fix(X, Y, Z, regr.meth = "gam", weight.num = 7, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Z, regr.meth = "xgboost", weight.num = 7, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Zm, regr.meth = "gam", weight.num = 7, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Zm, regr.meth = "xgboost", weight.num = 7, weight.meth = "sign"))

  #weight.num = 0
  expect_silent(wgcm.fix(X, Y, Z, regr.meth = "gam", weight.num = 0, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Z, regr.meth = "xgboost", weight.num = 0, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Zm, regr.meth = "gam", weight.num = 0, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Zm, regr.meth = "xgboost", weight.num = 0, weight.meth = "sign"))

  ##multivariate X and Y
  X <- cbind(X, X+rnorm(n), rnorm(n))
  Y <- cbind(Y, rnorm(n))
  #weight.num = 7
  expect_silent(wgcm.fix(X, Y, Z, regr.meth = "gam", weight.num = 7, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Z, regr.meth = "xgboost", weight.num = 7, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Zm, regr.meth = "gam", weight.num = 7, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Zm, regr.meth = "xgboost", weight.num = 7, weight.meth = "sign"))
  #weight.num = 0
  expect_silent(wgcm.fix(X, Y, Z, regr.meth = "gam", weight.num = 0, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Z, regr.meth = "xgboost", weight.num = 0, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Zm, regr.meth = "gam", weight.num = 0, weight.meth = "sign"))
  expect_silent(wgcm.fix(X, Y, Zm, regr.meth = "xgboost", weight.num = 0, weight.meth = "sign"))
})





R/utils.R
## function for multivariate wgcm.fix
wgcm.fix.mult <- function(X, Y, Z, regr.meth, regr.pars, weight.num,
                          weight.meth, nsim){
  X <- as.matrix(X)
  Y <- as.matrix(Y)
  n <- NROW(X)
  dx <- NCOL(X)
  dy <- NCOL(Y)
  if(is.null(Z)){
    warning("No Z specified. No weight functions can be calculated. Function simply tests for vanishing correlation between components of X and Y.")
    eps.mat <- t(t(X)-colMeans(X))
    xi.mat <- t(t(Y)-colMeans(Y))
    W <- rep(1,n)
  } else {
    Z <- as.matrix(Z)
    calc.res.Z <- function(V){
      return(as.numeric(GeneralisedCovarianceMeasure::comp.resids(V, Z,
                                                                  regr.pars = regr.pars,
                                                                  regr.method = regr.meth)))
    }
    eps.mat <- apply(X,2, calc.res.Z)
    xi.mat <- apply(Y,2,calc.res.Z)
    W <- weight_matrix(Z, weight.num, weight.meth)
  }
  R <- NULL
  for (j in 1:dx){
    for (l in 1:dy){
      R.jl <- eps.mat[,j]*xi.mat[,l]
      R <- cbind(R, R.jl*W)
    }
  }
  R <- t(R)
  R.norm <- R / sqrt(rowMeans(R^2) - rowMeans(R)^2)
  T.stat <- sqrt(n) * max(abs(rowMeans(R.norm)))
  T.stat.sim <- apply(abs(R.norm %*% matrix(rnorm(n * nsim), n, nsim)),
                      2, max) / sqrt(n)
  p.value <- (sum(T.stat.sim >= T.stat) + 1) / (nsim + 1)
  return(p.value)
}

## function for multivariate wgcm.est
wgcm.est.mult <- function(X, Y, Z, beta, regr.meth, regr.pars, nsim){
  X <- as.matrix(X)
  Y <- as.matrix(Y)
  n <- NROW(X)
  dx <- NCOL(X)
  dy <- NCOL(Y)
  if(is.null(Z)){
    warning("No Z specified. No weight functions can be estimated. Function simply tests for vanishing correlation between components of X and Y.")
    eps.mat <- t(t(X)-colMeans(X))
    xi.mat <- t(t(Y)-colMeans(Y))
    R <- NULL
    for (j in 1:dx){
      for (l in 1:dy){
        R.jl <- eps.mat[,j]*xi.mat[,l]
        R <- cbind(R, R.jl)
      }
    }
    R <- t(R)
    R.norm <- R / sqrt(rowMeans(R^2) - rowMeans(R)^2)
    T.stat <- sqrt(n) * max(abs(rowMeans(R.norm)))
    T.stat.sim <- apply(abs(R.norm %*% matrix(rnorm(n * nsim), n, nsim)),
                        2, max) / sqrt(n)
    p.value <- (sum(T.stat.sim >= T.stat) + 1) / (nsim + 1)
  } else {
    Z <- as.matrix(Z)
    ind.train <- sample(1:n, ceiling(beta*n))
    Xtrain <- X[ind.train,]
    Xtest <- X[-ind.train,]
    Ytrain <- Y[ind.train,]
    Ytest <- Y[-ind.train,]
    Ztrain <- Z[ind.train,]
    Ztest <- Z[-ind.train,]
    calc.res.Z <- function(V){
      return(as.numeric(GeneralisedCovarianceMeasure::comp.resids(V, Ztest,
                                                                  regr.pars = regr.pars,
                                                                  regr.method = regr.meth)))
    }
    eps.mat <- apply(Xtest, 2, calc.res.Z)
    xi.mat <- apply(Ytest, 2, calc.res.Z)
    R <- NULL
    for (j in 1:dx){
      for (l in 1:dy){
        W <- predict_weight(Xtrain[,j], Ytrain[,l], Ztrain, Ztest, regr.meth, regr.pars)
        R.jl <- eps.mat[,j]*xi.mat[,l]
        R <- cbind(R, R.jl*W)
      }
    }
    R <- t(R)
    R.norm <- R/sqrt(rowMeans(R^2)-rowMeans(R)^2)
    ntest <- n-length(ind.train)
    #The estimated w-functions aim at making a positive test statistic. Hence
    #a one-sided test.
    T.stat <- sqrt(ntest) * max(rowMeans(R.norm))
    T.stat.sim <- apply(R.norm %*% matrix(rnorm(ntest * nsim), ntest, nsim),
                        2, max) / sqrt(ntest)
    p.value <- (sum(T.stat.sim >= T.stat) + 1)/(nsim + 1)
  }
  return(p.value)
}



## function to calculate weight matrix
weight_matrix <- function(Z, weight.num, weight.meth) {
  if (weight.meth == "sign") {
    n <- NROW(Z)
    dz <- NCOL(Z)
    W <- rep(1,n)
    if (weight.num >= 1) {
      d.probs <- (1:weight.num) / (weight.num + 1)
      for (i in 1:dz) {
        Zi <- Z[,i]
        a.vec <- quantile(Zi, d.probs, names=F)
        Wi <- outer(Zi, a.vec, signa)
        W <- cbind(W,Wi)
      }
    }
  } else {
    stop("Only method \"sign\" implemented yet to calculate weight function")
  }
  return(W)
}




## translated sign weight function
signa <- function(x,a){return(sign(x-a))}


## function to calculate a 1sided p-value for wgcm.fix, since we expect
## the test statistic to be positive under the alternative
wgcm.1d.1sided <- function(Xtest, Ytest, Ztest, W, regr.meth, regr.pars) {
  n <- NROW(Ztest)
  eps <- as.numeric(GeneralisedCovarianceMeasure::comp.resids(Xtest, Ztest,
                                                              regr.pars = regr.pars, regr.method = regr.meth))
  xi <- as.numeric(GeneralisedCovarianceMeasure::comp.resids(Ytest, Ztest,
                                                             regr.pars = regr.pars, regr.method = regr.meth))
  R <- eps*xi*W
  T.stat <- sqrt(n)*mean(R)/sqrt(mean(R^2)-mean(R)^2)
  p.value <- 1-pnorm(T.stat)
  return(p.value)
}

## function to estimate weight function for wgcm.est
predict_weight <- function(Xtrain, Ytrain, Ztrain, Ztest,
                           regr.meth, regr.pars) {
  eps <- as.numeric(GeneralisedCovarianceMeasure::comp.resids(Xtrain, Ztrain,
                                regr.pars = regr.pars, regr.method = regr.meth))
  xi <- as.numeric(GeneralisedCovarianceMeasure::comp.resids(Ytrain, Ztrain,
                                regr.pars = regr.pars, regr.method = regr.meth))
  eps.xi.train <- eps * xi
  switch(regr.meth, "gam"={
    W <- sign(predict_weight_gam(eps.xi.train, Ztrain, Ztest, regr.pars))
  }, "xgboost"={
    W <- sign(predict_weight_xgboost(eps.xi.train, Ztrain, Ztest, regr.pars))
  }
  )
  return(W)
}



## function to estimate weight function for wgcm.est using gam
predict_weight_gam <- function(V, Z, Ztest, pars = list()) {
  mod <- train.gam(Z, V, pars = list())
  dz <- NCOL(Z)
  dpred <- data.frame(Ztest)
  names(dpred) <- paste("var", 2:(dz+1), sep = "")
  Vpred <- predict(mod$model, dpred)
  return(Vpred)
}

## function to estimate weight function for wgcm.est using xgboost
predict_weight_xgboost <- function(V, Z, Ztest, pars = list()) {
  mod <- train.xgboost1(Z, V, pars = list())
  dz <- NCOL(Z)
  dpred <- xgb.DMatrix(data = data.matrix(Ztest))
  Vpred <- predict(mod$model, dpred)
  return(Vpred)
}




R/trainFunctions.R
#The code in trainFunctions.R is copied (with two small modifications in the
#second function) from the R package 'GeneralisedCovarianceMeasure' by
#Jonas Peters and Rajen D. Shah released on CRAN
#<https://cran.r-project.org/package=GeneralisedCovarianceMeasure>
#under the GPL-2 license. Jonas Peters and Rajen D. Shah are the authors and
#copyright holders of the following two functions.


train.gam <- function (X, y, pars = list()) {
  if (!exists("numBasisFcts", pars)) {
    pars$numBasisFcts <- 100
  }
  if (!exists("staysilent", pars)) {
    pars$staysilent <- TRUE
  }
  if (!exists("CV.folds", pars)) {
    pars$CV.folds <- NA
  }
  if (is.null(X) || dim(as.matrix(X))[2] == 0) {
    result <- list()
    result$Yfit <- as.matrix(rep(mean(y), length(y)))
    result$residuals <- as.matrix(y - result$Yfit)
    result$model <- NA
    result$df <- NA
    result$edf <- NA
    result$edf1 <- NA
    result$p.values <- NA
  }
  else {
    p <- dim(as.matrix(X))
    if (!is.na(pars$CV.folds)) {
      num.folds <- pars$CV.folds
      rmse <- Inf
      whichfold <- sample(rep(1:num.folds, length.out = p[1]))
      for (j in 1:length(pars$numBasisFcts)) {
        mod <- train.gam(as.matrix(X)[whichfold == j,
        ], y[whichfold == j], pars = list(numBasisFcts = pars$numBasisFcts,
                                          CV.folds = NA))
        datframe <- data.frame(as.matrix(X)[whichfold !=
                                              j, ])
        names(datframe) <- paste("var", 2:p[2],
                                 sep = "")
        rmse.tmp <- sum((predict(mod$model, datframe) -
                           y[whichfold != j])^2)
        if (rmse.tmp < rmse) {
          rmse <- rmse.tmp
          final.numBasisFcts <- pars$numBasisFcts[j]
        }
      }
    }
    else {
      final.numBasisFcts <- pars$numBasisFcts
    }
    if (p[1]/p[2] < 3 * final.numBasisFcts) {
      final.numBasisFcts <- ceiling(p[1]/(3 * p[2]))
      if (pars$staysilent == FALSE) {
        cat("changed number of basis functions to    ",
            final.numBasisFcts, "    in order to have enough samples per basis function\n")
      }
    }
    dat <- data.frame(as.matrix(y), as.matrix(X))
    coln <- rep("null", p[2] + 1)
    for (i in 1:(p[2] + 1)) {
      coln[i] <- paste("var", i, sep = "")
    }
    colnames(dat) <- coln
    labs <- "var1 ~ "
    if (p[2] > 1) {
      for (i in 2:p[2]) {
        labs <- paste(labs, "s(var", i, ",k = ",
                      final.numBasisFcts, ") + ", sep = "")
      }
    }
    labs <- paste(labs, "s(var", p[2] + 1, ",k = ",
                  final.numBasisFcts, ")", sep = "")
    mod_gam <- FALSE
    try(mod_gam <- gam(formula = formula(labs), data = dat),
        silent = TRUE)
    if (typeof(mod_gam) == "logical") {
      cat("There was some error with gam. The smoothing parameter is set to zero.\n")
      labs <- "var1 ~ "
      if (p[2] > 1) {
        for (i in 2:p[2]) {
          labs <- paste(labs, "s(var", i, ",k = ",
                        final.numBasisFcts, ",sp=0) + ", sep = "")
        }
      }
      labs <- paste(labs, "s(var", p[2] + 1, ",k = ",
                    final.numBasisFcts, ",sp=0)", sep = "")
      mod_gam <- gam(formula = formula(labs), data = dat)
    }
    result <- list()
    result$Yfit <- as.matrix(mod_gam$fitted.values)
    result$residuals <- as.matrix(mod_gam$residuals)
    result$model <- mod_gam
    result$df <- mod_gam$df.residual
    result$edf <- mod_gam$edf
    result$edf1 <- mod_gam$edf1
    result$p.values <- summary.gam(mod_gam)$s.pv
  }
  return(result)
}


train.xgboost1 <- function (X, y, pars = list()) {
  n <- length(y)
  if (!exists("nrounds", pars)) {
    pars$nrounds <- 50
  }
  if (!exists("max_depth", pars)) {
    pars$max_depth <- c(1, 3, 4, 5, 6)
  }
  if (!exists("CV.folds", pars)) {
    pars$CV.folds <- 10
  }
  if (!exists("ncores", pars)) {
    pars$ncores <- 1
  }
  if (!exists("early_stopping", pars)) {
    pars$early_stopping <- 10
  }
  if (!exists("silent", pars)) {
    pars$silent <- TRUE
  }
  if (is.null(X) || dim(as.matrix(X))[2] == 0) {
    result <- list()
    result$Yfit <- as.matrix(rep(mean(y), length(y)))
    result$residuals <- as.matrix(y - result$Yfit)
    result$model <- NA
    result$df <- NA
    result$edf <- NA
    result$edf1 <- NA
    result$p.values <- NA
  }
  else {
    X <- as.matrix(X)
    if (!is.na(pars$CV.folds)) {
      num.folds <- pars$CV.folds
      rmse <- matrix(0, pars$nrounds, length(pars$max_depth))
      set.seed(1)
      whichfold <- sample(rep(1:num.folds, length.out = n))
      for (j in 1:length(pars$max_depth)) {
        max_depth <- pars$max_depth[j]
        for (i in 1:10) {
          dtrain <- xgb.DMatrix(data = data.matrix(X[whichfold !=
                                                       i, ]), label = y[whichfold != i])
          dtest <- xgb.DMatrix(data = data.matrix(X[whichfold ==
                                                      i, ]), label = y[whichfold == i])
          watchlist <- list(train = dtrain, test = dtest)
          if (pars$ncores > 1) {
            bst <- xgb.train(data = dtrain, nthread = pars$ncores,
                             watchlist = watchlist, nrounds = pars$nrounds,
                             max_depth = max_depth, verbose = FALSE,
                             early_stopping_rounds = pars$early_stopping,
                             callbacks = list(cb.evaluation.log()))
          }
          else {
            bst <- xgb.train(data = dtrain, nthread = 1,
                             watchlist = watchlist, nrounds = pars$nrounds,
                             max_depth = max_depth, verbose = FALSE,
                             early_stopping_rounds = pars$early_stopping,
                             callbacks = list(cb.evaluation.log()))
          }
          ## MODIFICATION BY CYRILL SCHEIDEGGER
          ########################################
          #newscore <- (bst$evaluation_log[,3])^1
          newscore <- bst$evaluation_log$test_rmse
          ########################################
          ## END OF MODIFICATION BY CYRILL SCHEIDEGGER
          if (length(newscore) < pars$nrounds) {
            newscore <- c(newscore, rep(Inf, pars$nrounds -
                                          length(newscore)))
          }
          rmse[, j] <- rmse[, j] + newscore
        }
      }
      mins <- arrayInd(which.min(rmse), .dim = dim(rmse))
      if (!pars$silent) {
        show(rmse)
        show(mins)
        if ((mins[1] == 1) | (mins[1] == pars$nrounds) |
            (mins[2] == 1) | (mins[2] == length(pars$max_depth))) {
          show("There have been parameters selected that were the most extreme of the CV values")
          show(mins)
        }
      }
      final.nrounds <- mins[1]
      final.max_depth <- pars$max_depth[mins[2]]
    }
    else {
      if (length(pars$max_depth) > 1) {
        stop("providing a vector of parameters must be used with CV")
      }
      final.max_depth <- pars$max_depth
      final.nrounds <- pars$nrounds
    }
    dtrain <- xgb.DMatrix(data = data.matrix(X), label = y)
    bstY <- xgb.train(data = dtrain, nrounds = final.nrounds,
                      max_depth = final.max_depth, verbose = !pars$silent)
    result <- list()
    result$Yfit <- predict(bstY, data.matrix(X))
    result$residuals <- as.matrix(y - result$Yfit)
    ## MODIFICATION BY CYRILL SCHEIDEGGER:
    ####################
    # result$model <- NA
    result$model <- bstY
    ####################
    ## END OF MODIFICATION BY CYRILL SCHEIDEGGER
    result$df <- NA
    result$edf <- NA
    result$edf1 <- NA
    result$p.values <- NA
  }
  return(result)
}







R/WGCM.R
#' Weighted Generalised Covariance Measure (WGCM) With Fixed Weight Functions
#' Conditional Independence Test
#'
#' The Weighted Generalised Covariance Measure (WGCM) with Fixed Weight Functions
#' is a test for conditional independence. It is a generalisation of the
#' Generalised Covariance Measure implemented in the R package
#' GeneralisedCovarianceMeasure.
#'
#' @param X A (n x d_X) numeric matrix with n observations of d_X variables.
#' @param Y A (n x d_Y) numeric matrix with n observations of d_Y variables.
#' @param Z A (n x d_Z) numeric matrix with n observations of d_Z variables.
#' @param regr.meth One of "gam" and "xgboost" indicating the regression method
#' used to estimate the conditional expectations E[X|Z] and E[Y|Z].
#' @param regr.pars Optional additional regression parameters according to
#' GeneralisedCovarianceMeasure::comp.resids().
#' @param weight.num Number k_0 of weight functions per dimension of Z to be
#' used additionally to the constant weight function w(z) = 1. The total number
#' of weight functions will be 1 + k_0 * d_Z. In case of max(d_X, d_Y) > 1, the
#' same 1 + k_0 * d_Z weight functions are used for every combination of the
#' components of X and Y.
#' @param weight.meth String indicating the method to choose the weight functions.
#' Currently, only "sign" is implemented.
#' @param nsim Number of samples used to calculate the p-value using simulation.
#'
#' @return A p-value for the null hypothesis of conditional independence of X
#' and Y given Z.
#'
#' @references Please cite the following papers.
#' Cyrill Scheidegger, Julia Hoerrmann, Peter Buehlmann:
#' "The Weighted Generalised Covariance Measure"
#' \url{https://arxiv.org/abs/2111.04361}
#'
#' Rajen D. Shah, Jonas Peters:
#' "The Hardness of Conditional Independence Testing and the Generalised
#' Covariance Measure"
#' \url{https://arxiv.org/abs/1804.07203}
#'
#' @examples
#' set.seed(1)
#' n <- 200
#' Z <- rnorm(n)
#' X <- Z + 0.3*rnorm(n)
#' Y1 <- Z + 0.3*rnorm(n)
#' Y2 <- Z + 0.3*rnorm(n) + 0.3*X
#' Y3 <- Z + 0.3*rnorm(n) + 0.15*X^2
#' wgcm.fix(X, Y1, Z, regr.meth = "gam", weight.num = 7, weight.meth = "sign")
#' wgcm.fix(X, Y2, Z, regr.meth = "gam", weight.num = 7, weight.meth = "sign")
#' wgcm.fix(X, Y3, Z, regr.meth = "gam", weight.num = 7, weight.meth = "sign")
#'
#' @export
#'
#' @importFrom stats pnorm quantile rnorm
wgcm.fix <- function(X, Y, Z, regr.meth, regr.pars = list(),
                     weight.num, weight.meth = "sign", nsim = 499) {
  if ((NCOL(X) != 1) || (NCOL(Y) != 1)) {
    p.value <- wgcm.fix.mult(X, Y, Z, regr.meth, regr.pars,
                             weight.num, weight.meth, nsim)
  } else{
    n <- NROW(X)
    if (is.null(Z)) {
      warning("No Z specified. No weight functions can be calculated. Function simply tests for vanishing correlation between X and Y.")
      weight.num <- 0
      eps <- X - mean(X)
      xi <- Y - mean(Y)
    } else {
      Z <- as.matrix(Z)
      eps <- as.numeric(GeneralisedCovarianceMeasure::comp.resids(X, Z,
                                                                  regr.pars = regr.pars, regr.method = regr.meth))
      xi <- as.numeric(GeneralisedCovarianceMeasure::comp.resids(Y, Z,
                                                                 regr.pars = regr.pars, regr.method = regr.meth))
    }
    if (weight.num == 0) {
      R <- eps * xi
      T.stat <- sqrt(n) * mean(R) / sqrt(mean(R^2) - mean(R)^2)
      p.value <- 2 * pnorm(-abs(T.stat))
    } else {
      W <- weight_matrix(Z, weight.num, weight.meth)
      R <- eps * xi * W
      R <- t(R)
      R.norm <- R / sqrt(rowMeans(R^2) - rowMeans(R)^2)
      T.stat <- sqrt(n) * max(abs(rowMeans(R.norm)))
      T.stat.sim <- apply(abs(R.norm %*% matrix(rnorm(n * nsim), n, nsim)),
                          2, max) / sqrt(n)
      p.value <- (sum(T.stat.sim >= T.stat) + 1) / (nsim + 1)
    }
  }
  return(p.value)
}

#' Weighted Generalised Covariance Measure (WGCM) With Estimated Weight Function
#' Conditional Independence Test
#'
#' The Weighted Generalised Covariance Measure (WGCM) with Estimated Weight Function
#' is a test for conditional independence. It is a generalisation of the
#' Generalised Covariance Measure implemented in the R package
#' GeneralisedCovarianceMeasure.
#'
#' @param X A (n x d_X) numeric matrix with n observations of d_X variables.
#' @param Y A (n x d_Y) numeric matrix with n observations of d_Y variables.
#' @param Z A (n x d_Z) numeric matrix with n observations of d_Z variables.
#' @param beta A real number between 0 and 1 indicating the fraction of the sample
#' used to estimate the weight function.
#' @param regr.meth One of "gam" and "xgboost" indicating the regression method
#' used to estimate the conditional expectations E[X|Z] and E[Y|Z] and the
#' weight function sign(E[(X-E[X|Z])(Y-E[Y|Z])|Z]).
#' @param regr.pars Optional additional regression parameters according to
#' GeneralisedCovarianceMeasure::comp.resids()
#' @param nsim Number of samples used to calculate the p-value using simulation.
#' Only used if max(d_X, d_Y) > 1.
#'
#' @return A p-value for the null hypothesis of conditional independence of X
#' and Y given Z.
#'
#' @references Please cite the following papers.
#' Cyrill Scheidegger, Julia Hoerrmann, Peter Buehlmann:
#' "The Weighted Generalised Covariance Measure"
#' \url{https://arxiv.org/abs/2111.04361}
#'
#' Rajen D. Shah, Jonas Peters:
#' "The Hardness of Conditional Independence Testing and the Generalised
#' Covariance Measure"
#' \url{https://arxiv.org/abs/1804.07203}
#'
#' @examples
#' set.seed(1)
#' n <- 200
#' Z <- rnorm(n)
#' X <- Z + 0.3*rnorm(n)
#' Y1 <- Z + 0.3*rnorm(n)
#' Y2 <- Z + 0.3*rnorm(n) + 0.3*X
#' Y3 <- Z + 0.3*rnorm(n) + 0.15*X^2
#' wgcm.est(X, Y1, Z, beta = 0.3, regr.meth = "gam")
#' wgcm.est(X, Y2, Z, beta = 0.3, regr.meth = "gam")
#' wgcm.est(X, Y3, Z, beta = 0.3, regr.meth = "gam")
#'
#' @export
#'
#' @import xgboost
#' @import mgcv
#' @importFrom methods show
#' @importFrom stats formula
#' @importFrom stats predict
wgcm.est <- function(X, Y, Z, beta = 0.3, regr.meth, regr.pars = list(), nsim=499) {
  if ((NCOL(X) != 1) || (NCOL(Y) != 1)) {
    p.value <- wgcm.est.mult(X, Y, Z, beta, regr.meth, regr.pars, nsim)
  } else {
    n <- NROW(X)
    if (is.null(Z)){
      warning("No Z specified. No weight function can be estimated. Function simply tests for vanishing correlation between X and Y.")
      eps <- X - mean(X)
      xi <- Y - mean(Y)
      R <- eps * xi
      T.stat <- sqrt(n) * mean(R) / sqrt(mean(R^2) - mean(R)^2)
      p.value <- 2 * pnorm(-abs(T.stat))
    } else {
      Z <- as.matrix(Z)
      ind.train <- sample(1:n, ceiling(beta * n))
      Ztrain <- Z[ind.train, ]
      Ztest <- Z[-ind.train, ]
      W <-  predict_weight(X[ind.train], Y[ind.train], Ztrain, Ztest,
                           regr.meth, regr.pars)
      p.value <- wgcm.1d.1sided(X[-ind.train],Y[-ind.train], Ztest, W,
                                regr.meth, regr.pars)
    }
  }
  return(p.value)
}
